docs/rfcs/021-autoscaling-compute-cache.md

```diff
@@ -0,0 +1,201 @@
+    | | +--------------------+ | |               |            |
+    | | |  Postgres          | | |               | Pageserver |
+    | | |                    | | |               |            |
+    | | |  ################  | | |               |            |
+    | | |  # shared_      #  | | |               +------------+
+    | | |  # buffers      #  | | |
+    | | |  ################  | | |
+    | | |                    | | |
+    | | |  ################  | | |
+    | | |  # Local FS     #  | | |
+    | | |  # cache        #  | | |
+    | | |  #              #  | | |
+    | | |  ################  | | |
+    | | |                    | | |
+    | | +--------------------+ | |
+    | +------------------------+ |
+    +----------------------------+
+
+Pros:
+
+- No PostgreSQL core changes required
+- Automatically takes advantage of local SSDs, not just memory
+
+Cons:
+
+- Needs explicit operations to change the size of the cache file
+- Need support for live migration of the filesystem
+
+
+Question:
+How is the page cache shared between the host kernel and the VMs? Does
+each VM maintain their own page cache? I think that depends on the
+filesystem and qemu driver we choose. If we use a raw block device and
+let the VM manage it, I believe the VM will maintain the page
+cache. But if we use a driver to access the host filesystem directly,
+or use something like NFS, I'm not sure.
+
+
+# Alternative 3: Host cache
+
+Implement a new host cache process that intercepts all the I/O
+requests from all VMs running on the host, and manages a local cache.
+Postgres can communicate with the host cache using TCP, or via custom
+virtio driver.
+
+    +----------------------------+
+    |      Host                  |
+    |                            |
+    |   ######################   |
+    |   #                    #   |
+    |   # shared host cache  #   |
+    |   #                    #   |
+    |   #                    #   |
+    |   #                    #   |
+    |   #                    #   |
+    |   ######################   |
+    |                            |
+    | +------------------------+ |
+    | |    Container VM        | |               +------------+
+    | | +--------------------+ | |               |            |
+    | | |  Postgres          | | |               | Pageserver |
+    | | |                    | | |               |            |
+    | | |  ################  | | |               |            |
+    | | |  # shared_      #  | | |               +------------+
+    | | |  # buffers      #  | | |
+    | | |  ################  | | |
+    | | |                    | | |
+    | | +--------------------+ | |
+    | +------------------------+ |
+    +----------------------------+
+
+Pros:
+- dynamic sharing between all tenants (one cache and eviction policy for all)
+- No PostgreSQL core changes required
+- Takes advantage of local SSDs, not just memory
+
+Cons:
+
+- Whole new component to write
```

Comment:
Some testing is needed, but my prediction that is request-response latency crossing VM boundary will be approx the same as with the pageserver in the same AZ. Localhost does 30k rps, local network usually 10-20k rps. Contrary to shared memory access that can give million rps.

Reply:
We can certainly make some prototype of third approach but it seems to me that current local Neon configuration when pageserver and compute are launched at the same node can be considered as such prototype. If all pages are cached in page server (it can be achieved by tuning page cache size) then pageserver will act almost as this shared cache server.
So I am +1 to Stas that we will not get good performance with this approach.

It is certainly possible to optimize channel of communication between this shared cache server and compute (non-locking ring buffer, BBQ, something like shmpipe...) but I am not sure about safety of such implementation.
Compute is not protected from different vulnerabilities. If compute has direct access to shared memory then ... how can we protect data of other tenants?

Also I am not quite agree with list of pros/cons.
Lets consider problem of migration.
As far as I understand VM is migrated together with local storage. Otherwise Postgres will just not work.
If so, local file cache will be migrated together and there will be no need to prewarm migrated compute.
I am not familiar with VM migration. But if it is something similar with I have done with Virtual Box/VMware:
1. Suspend VM
2. Copy image to another node
3. Adjust VM parameter
4. Start new VM from this image
then nothing special has to be done for VM migration in case of local file cache.
There can be only one problem - size of this local cache. If it is several Tb, then coping this data will take some time.
But if all this data is actually needed for normal work of compute, then it is still the most efficient way of node startup:
loading each page on demand will require much ore time (x10 or even x100).

In first case (dynamically extended shared buffers) content of shared buffers will be included in memory dump and transferred to new node automatically.

Concerning third solution (shared cache server) - it is not clear how to migrate content of shared buffers.
| Approach | dynamic shared buffers | local file cache | global cache server |
| ------------- | ------------------------------  | ------------------- | --------------------------- |
| Migration |  + | + | - |
| Complexity of implementation | high | low | very high | 
| Price/performance | 1 | 3 | 2 |  
| Flexibility | 1 | 3 | 2 |
| Resources utilization | 1 | 3  | 2 |

And my version of pros/cons:
Dynamic shred buffers
Pros:
     1. Maximal performance if result set fits in shared buffer
     2. Amazon Aurora has  it.
Cons:
     1. Limited by RAM size 
     2. Requires changes in Postgres core
     3. Slow reaction time (cache can not be shrinked immediately).         
     4. May require use of pg_prewarm to load this cache.
     
Local file cache:
Pros:
     1. Easy to implement
     2. SSD is less expensive than RAM
     3. Doesn't require explicit prewarming
     4. No changes in Postgres core
Cons:
      1. No easy way to share local disk space between tenants.
      2. Reading page from OS cache is slower then direct access to shared buffers
      3. Provides performance comparable with shared buffers only if all working set fits in local cache and in OS file cache.
      4. No easy way to preserve cache after node restart

Shared cache server:
Pros:
      1. Most efficient RAM usage at node
      2. No changes in Postgres core
Cons:
      1. Limited by RAM size
      2. Requires IPC which can be battlement for reaching good performance.
      3. Security challenges
      4. Problems with migration
      5. Most complex solution for implementation